exorcist/taskdb.py

```diff
@@ -1,40 +1,226 @@
+import abc
 import sqlalchemy as sqla
 
+# remaining imports are for typing
+from typing import Optional, Iterable
+from .models import TaskStatus, Task
+from os import PathLike
+import networkx as nx
 
-def create_task_db(engine):
+
+class NoStatusChange(Exception):
+    """Raised when an attempt to change task status does not change status.
+
+    In most cases, like when a task status is changed to "completed", if the
+    task is already marked as the new status, this indicates database
+    corruption.
+
+    Raising this exception has two purposes:
+
+    1. To enable it to be caught and allow the database corruption to be
+       fixed, if possible, and exit otherwise.
+    2. So that, in the event that the status that needs to be set on early
+       exit (e.g., due to a sigterm send by the queuing system) is the same
+       as the current status, it can be caught without error and exit
+       normally.
     """
-    Create an empty task database with the provided engine.
+
+
+class AbstractTaskStatusDB(abc.ABC):
+    @abc.abstractmethod
+    def add_task(self, taskid: str, requirements: Iterable[str]):
+        """Add a task to the database.
+
+        Parameters
+        ----------
+        taskid: str
+            the taskid to add to the database
+        requirements: Iterable[str]
+            taskids that directly block the task to be added (typically,
+            whose outputs are inputs to the task)
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def add_task_network(self, task_network: nx.DiGraph):
+        """Add a network of tasks to the database.
+
+        Parameters
+        ----------
+        task_network: nx.Digraph
+            A network with taskid strings as nodes.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def check_out_task(self) -> str:
+        """
+        Select a task to be run.
+
+        This should include any internal updates to the task status database
+        to indicate that the given task has been checked out by a worker.
+
+        Returns
+        -------
+        str :
+            The taskid of the task to run
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def mark_task_aborted_incomplete(self, taskid: str):
+        """
+        Update the database when a task fails to complete.
```

Comment:
it's not clear how this is different from completed with success=False.  Does this not increment the number of failures?  It just reverts the task to its status before it was checked out?

Reply:
See https://github.com/OpenFreeEnergy/exorcist/pull/9#issuecomment-1604727304

Reply:
So if that's the intent put it in the docstring, else it'll never happen.  If it's not thought out yet, don't include it yet.  We're going for minimal.